### ISSUE TITLE
Usage with Redux toolkits defaultMiddleware and configureStore()

### ISSUE THREAD

**Issue by MartinDawson** [dai-shi/redux-in-worker#11]
Hi,

First off thanks for the great library, it works really well.

My question is, how can we intregrate this with redux toolkits functions?

I'm following the pattern from your router example.

For example, I try something like this:

```js
    innerStore = configureStore({
        reducer,
        preloadedState: {
          ...initialState,
          ...preloadedState,
        },
        enhancers: [enhancer],
      });
```

And I get some obscure error. Also I tried with `getDefaultMiddelware` like so:

```js
  const outerStore = wrapStore(worker, undefined, compose(enhancer, ...getDefaultMiddleware()));
```

If you know how I can integrate these functions it would be great.

Thanks!

**Comment by dai-shi** [comment]
If I understand it correctly, RTK depends on redux-thunk, which is really bad at webworkers.
It's unfortunate to tell this, but it will never be possible with the current approach.

See my blog post too: https://blog.axlight.com/posts/redux-in-worker-off-main-thread-redux-reducers-and-middleware/

---
If you use RTK only in a worker side, it may work. But, that means you can't get the full benefit of RTK.

**Comment by MartinDawson** [comment]
Sorry let me rephrase. So I have two stores as per your router example, one for the web worker and one for the rest of my normal reducers.

Here's what I'm trying:

```js
 innerStore = createStore(
        reducer,
        {
          ...initialState,
          ...preloadedState,
        },
        compose(applyMiddleware(thunk), enhancer),
      );
```

```js
 const outerStore = wrapStore(
    worker,
    undefined,
    compose(applyMiddleware(thunk), enhancer, devTool),
  );
```

1. Is this the correct way to apply middleware to the enhancers? The outerStore one works fine but the innerStore one throws.
2. I don't actually want to use redux thunk with webworkers, only with the rest of my other non worker reducers. Or is this still not possible and I have to switch to sagas?

**Comment by dai-shi** [comment]
1: I guess the innerStore is a mirror of the store in web worker, so not sure if you can add an enhancer that is incompatible with it. (You can apply middleware in the worker though.)
2: Hm, the router example is a special one, but in general outer store and inner store abstraction would work. But they should probably isolated, because the outer store can't know when the inner store is modified without a further trick.

Maybe you want to start with a simpler case with outer/inner stores and see how it works. The enhancer part is just redux thing, and has nothing to do with this lib.

Feel free to ask questions about the expected behavior and the source code of this lib.

**Comment by MartinDawson** [comment]
@dai-shi Thank you for the quick replies. Refreshing to see such quick help on open source issues.

I figured out my issue. It works with redux thunks for the rest of my reducers which is great!

Solution here: https://stackoverflow.com/questions/57317603/what-is-the-correct-way-to-add-additional-middleware-through-a-redux-store-enhan

```js
  innerStore.dispatch = thunk(innerStore)(innerStore.dispatch);

  const outerStore = wrapStore(
    worker,
    undefined,
    compose(enhancer, applyMiddleware(thunk), devTool),
  );
```

Here's my full working code (messy) in case anyone else needs help applying middleware and getting this setup. It's quite tricky.

```js

const createStore = (preloadedState, reducers) => {
  let innerStore;

  const enhancer = (createStore) => {
    return (originalReducer, initialState, enhancer) => {
      const reducer = (state, action) => {
        const dcfState = originalReducer(undefined, action);
        const newState = {
          ...state,
          dcf: {
            ...state.dcf,
            ...dcfState,
          },
        };

        return combineReducers({
          fundamentals: fundamentalsReducer,
          dcf: (state = {}) => state,
          ...reducers,
        })(newState, action);
      };

      innerStore = createStore(
        reducer,
        {
          ...initialState,
          ...preloadedState,
        },
        enhancer,
      );

      innerStore.dispatch = thunk(innerStore)(innerStore.dispatch);

      return innerStore;
    };
  };

  let worker;
  let devTool;

  // For SSR
  if (typeof window !== "undefined") {
    worker = new Worker("../workers/dcfStore.worker.js", {
      type: "module",
    });
    devTool = window.__REDUX_DEVTOOLS_EXTENSION__
      ? window.__REDUX_DEVTOOLS_EXTENSION__()
      : (x) => x;
  } else {
    worker = {
      postmessage: () => {},
      onerror: () => {},
      onmessageerror: () => {},
    };
    devTool = (x) => x;
  }

  const outerStore = wrapStore(
    worker,
    undefined,
    compose(enhancer, applyMiddleware(thunk), devTool),
  );

  const dispatch = (action) => {
    outerStore.dispatch(action);
    innerStore.dispatch(action);

    return action;
  };

  const store = {
    ...outerStore,
    dispatch,
  };

  return store;
};
```

And my dcfStore.worker.js is using redux toolkits `createReducer`. It seems to work great!

Next up I will try to use redux toolkits `getDefaultMiddleware()`, but it should be the same principle. Just wrapping the dispatches on this part: `innerStore.dispatch = thunk(innerStore)(innerStore.dispatch);`
